### ISSUE TITLE
Run scheduler PostFilters asynchronously to improve throughput

### ISSUE THREAD

**Issue by mengqiy** [kubernetes/kubernetes#126858]
### What would you like to be added?

It should improve the scheduler throughput if we can
- run PostFilters asynchronously. 
- update current pod status asynchronously

### Why is this needed?

When a pod can't be scheduled, the scheduler may attempt to preempt other pods synchronously. The preemption requires up to 3 API calls (1. update pod condition 2. update pod status 3. preempt victim pod). This can slow down the scheduler unnecessarily when the APIServer has higher latency (e.g. under heavy load).
https://github.com/kubernetes/kubernetes/blob/44ce613173770aaeac6777a22059398096fb17c6/pkg/scheduler/framework/preemption/preemption.go#L372-L380
Given we only run one single `ScheduleOne` goroutine to schedule pods serially, we should make one single scheduling attempt as fast as possible.

Update current pod status when a pod is unschedulable is in the sync path too. We should change it to be async similarly as what we do with bind process.
https://github.com/kubernetes/kubernetes/blob/850bfd9aa92423696dbb6de2070ab7494fdd99b1/pkg/scheduler/schedule_one.go#L112

**Comment by mengqiy** [comment]
@kubernetes/sig-scheduling-misc Any thoughts?

**Comment by alculquicondor** [comment]
I wouldn't run all of PostFilter asynchronously, as it depends on what is actually scheduled.

However, once all the preemption victims are calculated, we can potentially send the API calls to preempt them asynchronously.

cc @macsko

**Comment by Huang-Wei** [comment]
> The preemption requires up to 3 API calls (1. update pod condition 2. update pod status 3. preempt victim pod).

It's two operations actually (the first [apipod.UpdatePodCondition()](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L274) just prepares an in-memory pod obj).

Function-wise, we need the aggregated result of victims' preemption (update victims' status and delete them) to decide whether or not proceed w/ the preemption cycle, so even if we made victims' preemption asynchrounous, we have to wait for the result, so I don't quite think we would benefit much. Also fwiw, the victims' preemption have been [parallelized](https://github.com/kubernetes/kubernetes/blob/44ce613173770aaeac6777a22059398096fb17c6/pkg/scheduler/framework/preemption/preemption.go#L391).

**Comment by alculquicondor** [comment]
> we have to wait for the result

Can you remind me why is that?

**Comment by Huang-Wei** [comment]
It's b/c if some victim's preemption failed (like deletion failed somehow), it's inappropriate to continue set .status.nominatedNodeName to the preemptor pod; in this case instead, we should abort the preemption cycle and ensure the preemptor can start its next _fresh_ scheduling cycle.

**Comment by hakuna-matatah** [comment]
My understanding is that performing post-filter (preemption) asynchronously could delay the scheduling of high-priority pods, as they may be re-queued with a back-off strategy , causing them to wait for another scheduling cycle( In cases if/when we don't want to wait for the preemption result). In contrast, synchronous preemption `might` immediately create room on a node, allowing the high-priority pod to be scheduled within the same cycle, a benefit that could be lost with asynchronous handling.

Is this correct understanding ?

**Comment by Huang-Wei** [comment]
> My understanding is that performing post-filter (preemption) asynchronously could delay the scheduling of high-priority pods

It's more on accuracy. If we have multiple in-flight pods experiencing post-filter logic **at the same time**, the preemption decision is based on their own snapshot (cluster snapshot is based on the moment the pod ends a regular scheduling cycle), and thus the proposed nominatedNode for a preemptor could be in conflict with others, which would cause further rounds of chaos back and forth, and very likely introduce regressions.

**Comment by sanposhiho** [comment]
Does it work? -
1. PostFilter calculates and decides which Pods to preempt.
2. PostFilter starts goroutine that runs preemption --(if all success)-> sets `.status.nominatedNodeName`.
3. Preemption plugin gates the Pod that the preemption is underway until the goroutine (2) is done.
4. When the goroutine (2) is done, un-gate the Pod. We may need an API from the queue to the plugin to wake up specific Pods. (like `PodsToActivate`)

So, the idea is waiting for the preemption process to be done within the queue, not waiting at the scheduling cycle.

**Comment by Huang-Wei** [comment]
> 2\. PostFilter starts goroutine for API calls; runs preemption --(if all success)-> sets `.status.nominatedNodeName`. If not success, unnominate the Pod in the scheduler cache.

What if two goroutines chose the same node to run preemption at the same time?

**Comment by sanposhiho** [comment]
In short, I believe we don't have to worry about that scenario. (unless I missed any other scenario..)

Let's consider the simple scenario where the cluster is super packed and we have two unsched pods.

In this case,
- The current scheduler handles the preemption for pod1, preempt all victim pods on node1, and the scheduler starts to schedule pod2.
- My scheduler (= the scheduler after my proposal) starts to schedule pod2 while the preemption is on-going on Node1 for pod1.

And, we have to consider these two scenarios

- pod2 is higher priority than pod1
- pod2 is equal or lower priority than pod1

In the case of the first scenario (pod2 is higher priority than pod1), the current scheduler allows pod2 to take node1, where the preemption for pod1 made the space. And, when pod1 comes back to the scheduling cycle, it (likely) cannot land on node1 because pod2 is scheduled there now.
And, the final result from my scheduler would be the same as the current scheduler; even if two preemptions for pod1 and pod2 select the same Node and same victims, when pod1 comes back to the schedule cycle, it (likely) cannot land on node1 because pod2 is nominated there now.

In the case of the second scenario (pod2 is equal or lower priority than pod1), pod2 should not take a place for pod1.
The current scheduler rejects pod2 because pod1 is nominated on node1, and pod2 will go through the preemption process while taking pod1 into consideration (the preemption takes the nominated Pods into consideration). 
When pod1 comes back to the scheduling cycle, it can land on node1.
And, if two preemptions for pod1 and pod2 select the same Node and same victims, the preemption for pod2 will make the space for pod1 and pod2 because, again, the preemption takes the nominated Pods into consideration.
When pod1 comes back to the scheduling cycle, it can land on node1.

So, after all, the current scheduler and my scheduler makes the same result in all cases even if two preemptions select the same node.

**Comment by alculquicondor** [comment]
IIRC, we don't keep an in-memory record of which pods are preempted, we just rely on the Pod informers to communicate that information back. We only keep an in-memory record about which Node the Pod is nominated to schedule on.

I think what @sanposhiho is proposing is roughly equivalent to what we have today. The likelihood of multiple Pods trying to preempt the same Pods will increase, but they should be able to resolve that (that is, preempt more pods) in subsequent cycles.

The worst case scenario is if pod2 no longer fits and it had previously chosen to preempt different pods than pod1. In this case, those were unnecessary preemptions.

I think its an acceptable tradeoff, although maybe a bit risky, so I would put a feature gate to it (so we need an accompanying KEP).

But before we do that, WDYT @Huang-Wei?

**Comment by hakuna-matatah** [comment]
>My scheduler (= the scheduler after my proposal) starts to schedule pod2 while the preemption is on-going on Node1 for pod1.

If we adopt @sanposhiho's proposed approach, there's a potential edge case in scenario 2: Suppose the new-scheduler is running a go routine to evict pods and create space. After eviction but just before nomination, pod2(lower-priority) could get nominated on that node because there is room. Since pod1 has higher priority, pod2 might need to be preempted immediately after nomination ? That could be unnecessary work ? Could this situation arise if preemption and nomination logic run in a goroutine while the queue is being processed? 

Am i missing anything here ? @sanposhiho @alculquicondor @Huang-Wei

**Comment by sanposhiho** [comment]
> Suppose the new-scheduler is running a go routine to evict pods and create space. After eviction but just before nomination, pod2(lower-priority) could get nominated on that node because there is room.

My proposal is nominating pods in the scheduler's cache before starting goroutine.
Meaning, pod1 is nominated before the goroutine for the preemption for pod1 is started.
So, pod2's preemption always treats pod1 like running on the node1 already.

**Comment by sanposhiho** [comment]
DRA also uses API calls in PostFilter, which we may also want to solve altogether.

https://github.com/kubernetes/kubernetes/blob/99a8a6fe258a456d51dddd3dec16d168d94e4eb1/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L1189

Another additional idea here though, would it be over-engineering if we create another extension point after PostFilter? 
Let's call it **PostPostFilter** for now 😅.

So, we define PostFilter plugin shouldn't make any API calls, and PostPostFilter plugin should do instead.
We run PostFilter synchronously but PostPostFilter plugin asynchronously. 

The flow would be very similar to what I proposed above, so it's just how we design it.

1. At PostFilter extension point, the preemption plugin decides which Pods to preempt, stores them in the cyclestate, and nominates the Pod to the Node.
2. The scheduler starts goroutine for PostPostFilter plugins and moves the Pod back to the unschedulableQ. The Pod is gated until PostPostFilter plugins are done. (Then, the scheduler can start next scheduling cycle for another pod.)
5. At PostPostFilter extension point, the preemption plugin runs the preemption actually.
6. When all PostPostFilter plugins are done, the scheduler un-gates the pod and moves it from unschedQ to activeQ/backoffQ.

Then, I believe DRA can move their PostFilter to PostPostFilter as well to get the same optimization; make pods wait for the API calls in the queue, not in the scheduling cycles.

**Comment by Huang-Wei** [comment]
> IIRC, we don't keep an in-memory record of which pods are preempted

@alculquicondor but this is what @sanposhiho 's proposal is based on.

> So, after all, the current scheduler and my scheduler makes the same result in all cases even if two preemptions select the same node.

@sanposhiho It's actually not the same case. The source of truth of a scheduling cycle is based on the cluster snapshot, so in your proposal - even we "hot" patched the in-memory cache, it won't change the scheduling/preemption decision. If you spot any logic (in a scheduling cycle) that reads from cache instead of snapshot, that is indeed a bug as snapshot info is the cornerstone each scheduling cycle relies on.

**Comment by sanposhiho** [comment]
I don't see your point. My proposal is nominating pods in the scheduler's cache **before** starting goroutine, and it does not require any hot patches.

So, the flow is:
pod1 is nominated to node1 in the cache (but the API calls are done later asynchronously) -> the scheduling cycle for pod2 is started. -> the pod2 reaches PostFilter.

Given the pod1 is nominated before the start of the second scheduling cycle, the pod2's scheduling and preemption treats the pod1 like running on node1.

**Comment by sanposhiho** [comment]
> IIRC, we don't keep an in-memory record of which pods are preempted
>> @alculquicondor but this is what @sanposhiho 's proposal is based on.

Regarding "nominating pods in the scheduler's cache" that I mentioned several times, I meant [AddNominatedPod](https://github.com/kubernetes/kubernetes/blob/605e94f6dfa9e98831e7ed0932dd18cd75d99e46/pkg/scheduler/internal/queue/nominator.go#L67). We can directly call it from the preemption plugin.

**Comment by Huang-Wei** [comment]
> pod1 is nominated to node1 in the cache (but the API calls are done later asynchronously) -> the scheduling cycle for pod2 is started. -> the pod2 reaches PostFilter.

ok, I got it - in your proposal, the logic of node nomination for a preemptor is racing-free; in other words, the node nomination occurs before any pods are deleted.

I think in a happy path - where victims' deletion are successful - this proposal should work. But upon a failure of victims' deletion, we have to invalidate the cache at the end of the goroutine; that'd mean a concurrent scheduling decision may carry a dirty snapshot and hence cannot make an accurate scheduling decision (and this inaccurate decision may continue in next cycles).
Here is a concrete example: pod1 holds a higher priority than pod2, and pod1 is nominated to nodeX, and afterwards starts a goroutine to delete victims; meanwhile pod2 enters its scheduling cycle, and b/c a high-priority pod1 is assumed to nominate on nodeX, it cannot be scheduled. But the fact is, in the aforementioned goroutine, the victims deletion may fail, and thus pod1 has no chance to be scheduled to nodeX, and that cache is invalidated at the end of the victims deletion goroutine. In the next scheduling cycle of pod1 and pod2, the situation may remain.

I'm not saying it's a bad pattern, actually scheduler is already using this "assume pod's cache -> async goroutine to finish a  job" pattern in the {scheduling cycle, binding cycle} paradigm. When using this pattern, the key ask is: 1) how **frequent** the async goroutine's logic may fail? 2) how **long** the async goroutine would run to have to lock a might-be-inaccurate cache? For {scheduling cycle, binding cycle}, a binding API request has rare chance to fail and the duration is short. While in the case this issue is discussing, IMHO we don't have clear answers for both questions. If the goroutine fail and hence we have to make up (invalidate the cache), it may cause further rounds of churns, and in worse case, may yield totally different behavior with current scheduling (e.g., pod2 and even more pods cannot be scheduled), which is the thing most concerning me.

**Comment by sanposhiho** [comment]
Yes, that is true. My proposal assumes that the preemption API calls are likely successful, and if actually not, then some scheduling cycles could make non-best scheduling results because of an assumed Pod on the Node.

But, I believe it's an acceptable downside. We can implement the metrics to see two points you mentioned, the success rate of preemption goroutines and the duration metric and the duration, during the alpha/beta period, and confirm whether we should really be concerned about the failure scenario that much.

**Comment by Huang-Wei** [comment]
@mengqiy in your prod usecase, do you have some rough metrics of avg len(victims) and deletion duration upon preemption? It would help sig-scheduling better evaluate this issue.

**Comment by sanposhiho** [comment]
@mengqiy 
FYI, the metrics are:
- `preemption_victims`
- `framework_extension_point_duration_seconds` with `extension_point=PostFilter` (or `plugin_execution_duration_seconds` with `extension_point=PostFilter` and `plugin=DefaultPreemption`) 
  - It includes the time that it takes to calculate the victims etc though, we don't have a duration metric that only includes API calls.

**Comment by sanposhiho** [comment]
/assign

Maybe we start KEP anyway to try to make it in this release (alpha)?

**Comment by sanposhiho** [comment]
Started a KEP https://github.com/kubernetes/enhancements/pull/4833 to continue the discussion, while everyone is on the same understanding about what the proposal proposes. Please provide a feedback on the PR! 
We're still in the early week of v1.32 release cycle, and hopefully, we can make it in v1.32 as alpha.

**Comment by JasonChen57** [comment]
Hey @sanposhiho, thanks so much for starting the KEP, very excited for this potential enhancement!

One question that I had: As @mengqiy noted in the original post, there's a blocking call to update the pod status for all unschedulable pods (not just those that trigger preemptions). Two relevant code pointers:
https://github.com/kubernetes/kubernetes/blob/850bfd9aa92423696dbb6de2070ab7494fdd99b1/pkg/scheduler/schedule_one.go#L112
https://github.com/kubernetes/kubernetes/blob/c3ebd95c837a21c0c79ab64721c28bba44331bf9/pkg/scheduler/schedule_one.go#L1133

This is relevant to pods that trigger preemptions (because it's how the nominated node names gets set for the pod) but is relevant to unschedulable pods that don't trigger preemptions as well (because it's how the pod's condition is updated). 

Would the above KEP be able to make this call non-blocking for all unschedulable pods? Or would that be out of scope and potentially a separate enhancement?

**Comment by sanposhiho** [comment]
@JasonChen57 
The KEP is only talking about the preemption plugin and that point isn't in the scope right now.
But, that one looks much easier to resolve; we can just ship a goroutine to patch pod status and then requeue the pod.

```go
go func() {
	updatePod(...)

	// We requeue the pod back to the queue after the update is done so that this pod won't be retried until getting the update.
	sched.SchedulingQueue.AddUnschedulableIfNotPresent(...)
}
```

@alculquicondor @Huang-Wei 
Is it safe enough to just make ↑ this change without the feature gate? Or should we include this enhancement in the KEP (= under the feature gate) too?

**Comment by alculquicondor** [comment]
Also, we should devise some perf tests that show the improvements. I think we already have some for unschedulable pods and preemptions.

@macsko can you share some details about them?

**Comment by sanposhiho** [comment]
> Aren't there implications to the inflight pods map as well? Anything asynchronous will always bring some risk IMO. 

Nothing to worry around the in-flight pods. It can already handle such Pods being handled concurrently, like binding cycles.
But, yeah, we cannot say the risk is zero of course.

> what are the use cases we are trying to optimize and why?

All unschedulable Pods go thru this PATCH API call, and we can save the time for API call with goroutine. But, we don't know how much this time actually is, so -

> Also, we should devise some perf tests that show the improvements. 

yep, this is true.

**Comment by macsko** [comment]
> I think we already have some for unschedulable pods and preemptions.

For preemption we have only the simple case: fill the nodes with low priority pods and then create a bunch of high priority pods. Then we get a scheduling blocked for some time until the preemptions happen. 

I think it would be good to add an another test case where we'll schedule both unschedulable pods that require preemption and directly schedulable pods in the second step to see how the synchronous (or new asynchronous) preemption will limit the scheduling throughput of directly schedulable pods. 

For unschedulable pods test case it's also very simple scenario: fill the nodes with pods and then try to schedule more.

**Comment by alculquicondor** [comment]
What is this test doing? https://github.com/kubernetes/kubernetes/blob/e5dd48efd07e8a052604b3073e0fafe7361ca689/test/integration/scheduler_perf/config/performance-config.yaml#L691

We should probably add some comments to describe some scenarios.

In any case, I would like to hear from @mengqiy which kind of scenarios he is thinking of. Both for preemption and "unschedulable".

BTW, having lots of unschedulable pods has never been advised in kube-scheduler. That's why we introduced "scheduling gates" and kueue.

**Comment by JasonChen57** [comment]
I'm not familiar with the performance tests, but if I'm understanding https://github.com/kubernetes/kubernetes/blob/e5dd48efd07e8a052604b3073e0fafe7361ca689/test/integration/scheduler_perf/config/performance-config.yaml#L691 correctly it measures the performance when we're scheduling schedulable pods (`pod-default.yaml`) but there's an initial existing set of unschedulable pods (`pod-large-cpu.yaml`).

And I think the test we would like to add is one where we measure performance when it comes to the scheduler's processing of unschedulable pods.

Regarding the case where we're handling unschedulable pods and not triggering preemption: in case it's helpful, an example of a scenario we'd like to avoid is scheduler starvation, where the scheduler is spending all of its time reprocessing a set of higher-priority unschedulable pods that it doesn't process lower-priority schedulable pods. The number of higher-priority unschedulable pods we can have before this happens is tied to the rate at which the scheduler is able to process unschedulable pods, which would be increased by making the PATCH API call non-blocking.

Granted, there would be other solutions to this as well, like avoiding having so many unschedulable pods as you mentioned, or perhaps adjusting `podMaxBackoffSeconds`. But making the PATCH API call non-blocking would help significantly.

**Comment by alculquicondor** [comment]
Just to add to the scenario, what do you think is a good reason for the pods to be unschedulable? Just capacity? I ask this because certain scheduling filters are slower than others, and they are sorted in such a way that we can break on the faster filters first.

**Comment by JasonChen57** [comment]
Unschedulable pods due to capacity was the main scenario I was thinking about. Although from what I've observed for us, the blocking PATCH API call is the biggest bottleneck, as compared to the time spent filtering.

**Comment by dom4ha** [comment]
I can devote some time to add more perf tests.

@alculquicondor shall they cover scheduling unschedulable and PATCH API, or more to measure the performance of other plugins/scenarios? Do you think the performance setup is representative enough to make such measures?

**Comment by alculquicondor** [comment]
The perf tests run with a real apiserver, so the measurements are, at least, correlated with E2E performance.

**Comment by dom4ha** [comment]
I'm adding the tests for preemption and unschedulable in https://github.com/kubernetes/kubernetes/pull/127829, so feel free to comment there if other scenarios are needed.

**Comment by sanposhiho** [comment]
Closing this issue as we implemented the async preemption as an alpha feature and the enhancement is tracked at:
https://github.com/kubernetes/enhancements/issues/4832
